src/Oscoin/Crypto/Address/Internal.hs

```diff
@@ -0,0 +1,125 @@
+{-# LANGUAGE UndecidableInstances #-}
+module Oscoin.Crypto.Address.Internal where
+
+import           Oscoin.Prelude
+
+import           Oscoin.Configuration
+import           Oscoin.Crypto.Hash
+import           Oscoin.Crypto.PubKey
+
+import           Codec.CBOR.ByteArray (ByteArray)
+import           Data.Tagged
+
+{------------------------------------------------------------------------------
+  Types
+------------------------------------------------------------------------------}
+
+-- An 'Address' is broken down as follows:
+--
+-- * /5/ bytes for the version prefix, so allocated:
+-- | 00 | 00 | 00 | 00 | 00 |
+--   ^    ^     ^   ^    ^
+--   |    |     |   |    |_ reserved (unused)
+--   |    |     |   |_ reserved (unused)
+--   |    |     |_ address format (only cbor for now)
+--   |    |_ address type (mainnet, testnet, devnet, multisig, etc)
+--   |_  protocol version (bumped on hard forks)
+--
+-- * /23/ bytes for the CBOR-encoded address payload
+--
+-- * /6/ bytes for the checksum, computed from the <prefix + payload>.
+--
+data Address c = Address
+    { addressPrefix  :: AddressPrefix
+    , addressPayload :: AddressPayload c
+    } deriving (Eq, Show)
+
+data AddressPrefix = AddressPrefix
+    { protocolVersion :: ProtocolVersion
+    , addressType     :: AddressType
+    , addressFormat   :: AddressFormat
+    } deriving (Eq, Show)
+
+-- | A 'ProtocolVersion' is a byte representing the current version of the
+-- Oscoin protocol running. This makes possible to issue forks of the system
+-- by simply acting on this value, so that nodes running the old protocol
+-- won't be able to use addresses from the new fork. Conversely, nodes
+-- running the fork can be programmed to support both formats.
+newtype ProtocolVersion =
+    ProtocolVersion ProtocolVersionTag
+    deriving (Eq, Show)
+
+data ProtocolVersionTag =
+      ProtocolVersion_V1
+      deriving (Eq, Show)
+
+-- | The 'AddressType', i.e. the qualifying type of this 'Address'.
+-- Example:
+--
+-- 0x01 Mainnet
+-- 0x02 Testnet
+-- 0x03 Devnet
+-- ..
+newtype AddressType =
+    AddressType AddressTypeTag
+    deriving (Eq, Show)
+
+data AddressTypeTag =
+      AddressTypeMainnet -- (Mainnet, Production)
+    | AddressTypeTestNet -- (Testnet, Production)
+    | AddressTypeDevnet  -- (Devnet,  Production)
+    -- Other exotic permutations
+    | AddressTypeMainnetDev -- (Mainnet, Development)
+    | AddressTypeMainnetTst -- (Mainnet, Testing)
+    | AddressTypeTestNetDev -- (Testnet, Development)
+    | AddressTypeTestNetTst -- (Testnet, Testing)
+    | AddressTypeDevnetDev  -- (Devnet,  Development)
+    | AddressTypeDevnetTst  -- (Devnet,  Testing)
+    deriving (Eq, Show)
```

Comment:
Going through this extra type seems to add a lot of boilerplate, why not just always convert from `(network, env) -> Word8` and vice versa?

Reply:
We could -- however, I think this is exactly the boilerplate you want in these cases, i.e. this is a layer of indirection that allows you to not couple yourself with the (network, env) tuple just yet (in case that grows in the future). Plus, it's quite scary to have to deal directly with these kind of conversions, as it's quite easy to mistype something and end up with the wrong isomorphism (despite our roundtrip tests helps here, of course).

Reply:
84c5a8f